#  Pandas Tutorial

### Learn Pandas
* Pandas is a Python library.
* Pandas is used to analyze data.

### What is Pandas?
* `Pandas` is a Python library used for working with data sets.
* It has functions for *analyzing, cleaning, exploring, and manipulating data.*

In [60]:
import  pandas as pd

mydataset = {
  'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2]
}
 
myvar = pd.DataFrame(mydataset)
print(myvar)

    cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2


### Checking Pandas Version

In [61]:
print(pd.__version__)

2.2.3


### What is a Series?
* A Pandas Series is like a column in a table.
* It is a one-dimensional array holding data of any type.

In [65]:
a = [1, 2, 3, 4, 5]
myvar = pd.Series(a)
print(myvar)

0    1
1    2
2    3
3    4
4    5
dtype: int64


### Labels
* If nothing else is specified, the values are labeled with their index number. First value has index 0, second value has index 1 etc.

* This label can be used to access a specified value.

In [66]:
b = [6, 7, 8, 9, 10]
myvar = pd.Series(b, index=["x", "y", "z", "w", "v"])
print(myvar)

x     6
y     7
z     8
w     9
v    10
dtype: int64


### Key/Value Objects as Series
You can also use a `key/value object`, like a dictionary, when creating a Series.

In [67]:
calories = {
  "day1": 420,
  "day2": 380,
  "day3": 390
}
myvar = pd.Series(calories)
print(myvar)

day1    420
day2    380
day3    390
dtype: int64


In [69]:
# Create a Series using only data from "day1" and "day2":
calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories, index = ["day1", "day2"])

print(myvar)

day1    420
day2    380
dtype: int64


### DataFrames
* Data sets in Pandas are usually multi-dimensional tables, called DataFrames.
* `Series is like a column, a DataFrame is the whole table.`

In [71]:
import pandas as pd

data = {
  "calories": [420, 380, 390, 590],
  "duration": [50, 40, 45, 60]
}

myvar = pd.DataFrame(data)

print(myvar)

   calories  duration
0       420        50
1       380        40
2       390        45
3       590        60


## What is a DataFrame?
* A Pandas DataFrame is a `2 dimensional data structure`, like a 2 dimensional array, or a table with rows and columns.



In [76]:
import  pandas as pd

data = {
  "calories": [420, 380, 390, 590],
  "duration": [50, 40, 45, 60]
}
  
# Pandas use the loc attribute to return one or more specified row(s)  
df = pd.DataFrame(data)
print(df.loc[0])
print(df.loc[1:3])
print(df.loc[[0, 1, 2]])

calories    420
duration     50
Name: 0, dtype: int64
   calories  duration
1       380        40
2       390        45
3       590        60
   calories  duration
0       420        50
1       380        40
2       390        45


### Named Indexes
With the `index` argument, you can name your own indexes.

In [78]:
import pandas as pd
data = {
  "calories": [420, 380, 390, 590],
  "duration": [50, 40, 45, 60]
}
df = pd.DataFrame(data, index = ["day1", "day2", "day3", "day4"])
print(df)
print(df.loc["day2"])

      calories  duration
day1       420        50
day2       380        40
day3       390        45
day4       590        60
calories    380
duration     40
Name: day2, dtype: int64


## Pandas Read CSV

### Read CSV Files
* A simple way to store big data sets is to use `CSV files (comma separated files).`
* CSV files contains plain text and is a well know format that can be read by everyone including Pandas.

Tip: use `to_string()` to print the entire DataFrame.

In [80]:
import pandas as pd

df = pd.read_csv('data.csv')
print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

#### If you have a large DataFrame with many rows, Pandas will only return `the first 5 rows, and the last 5 rows:`

In [82]:
print(df)
print(pd.options.display.max_rows) 

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]
60


### max_rows
The number of rows returned is defined in Pandas option settings.
You can check your system's maximum rows with the `pd.options.display.max_rows` statement.

In [88]:
import pandas as pd

pd.options.display.max_rows = 9999

df = pd.read_csv('data.csv')

print(df) 

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

## Pandas Read JSON

## Read JSON
* Big data sets are often stored, or extracted as JSON.
* `JSON is plain text, but has the format of an object, and is well known in the world of programming, including Pandas.`

In [89]:
import pandas as pd

df = pd.read_json('https://www.w3schools.com/python/pandas/data.js')
print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

### Dictinonary as JSON 
* `JSON = Python Dictionary`

* JSON objects have the same format as Python dictionaries.

In [90]:
import pandas as pd

data = {
  "Duration":{
    "0":60,
    "1":60,
    "2":60,
    "3":45,
    "4":45,
    "5":60
  },
  "Pulse":{
    "0":110,
    "1":117,
    "2":103,
    "3":109,
    "4":117,
    "5":102
  },
  "Maxpulse":{
    "0":130,
    "1":145,
    "2":135,
    "3":175,
    "4":148,
    "5":127
  },
  "Calories":{
    "0":409,
    "1":479,
    "2":340,
    "3":282,
    "4":406,
    "5":300
  }
}

df = pd.DataFrame(data)

print(df) 

   Duration  Pulse  Maxpulse  Calories
0        60    110       130       409
1        60    117       145       479
2        60    103       135       340
3        45    109       175       282
4        45    117       148       406
5        60    102       127       300


# Pandas - Analyzing DataFrames

## Viewing the Data
* One of the most used method for getting a quick overview of the DataFrame, is the `head() method.`

* The `head() method` returns the headers and a specified number of rows, starting from the top.

In [98]:
import pandas as pd

df = pd.read_csv('data.csv')

print(df.head())
print("==" * 20)
print(df.head(10))


   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0
   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0
5        60    102       127     300.0
6        60    110       136     374.0
7        45    104       134     253.3
8        30    109       133     195.1
9        60     98       124     269.0


* There is also a `tail() method` for viewing the last rows of the DataFrame.
* The `tail() method` returns the headers and a specified number of rows, starting from the bottom.

In [99]:
print(df.tail())

     Duration  Pulse  Maxpulse  Calories
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4


### Info About the Data
* The DataFrames object has a method called `info(),` that gives you more information about the data set.

In [100]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB
None


### Null Values
* The info() method also tells us how many Non-Null values there are present in each column, and in our data set it seems like there are `164 of 169 Non-Null values in the "Calories" column.`

# Pandas - Cleaning Data

### Data Cleaning
Data cleaning means fixing bad data in your data set.

Bad data could be:
* Empty cells
* Data in wrong format
* Wrong data
* Duplicates

In [105]:
import pandas as pd

df = pd.read_csv('data1.csv')
# Display all data of the DataFrame
print(df.to_string())


    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'  

In [106]:
print("Dataset Information:")
df.info()

Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31 entries, 0 to 30
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  31 non-null     int64  
 1   Date      30 non-null     object 
 2   Pulse     31 non-null     int64  
 3   Maxpulse  31 non-null     int64  
 4   Calories  29 non-null     float64
dtypes: float64(1), int64(3), object(1)
memory usage: 1.3+ KB


## Pandas - Cleaning Empty Cells

### Empty Cells
* Empty cells can potentially give you a wrong result when you analyze data.

### Remove Rows
* One way to deal with empty cells is to remove rows that contain empty cells.

* This is usually OK, since data sets can be very big, and removing a few rows will not have a big impact on the result.



In [108]:
import pandas as pd

df = pd.read_csv("data1.csv")

new_df = df.dropna()
print("\nDataFrame after dropping rows with NaN values:")
print(new_df.to_string())


DataFrame after dropping rows with NaN values:
    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98 

`Note: By default, the dropna() method returns a new DataFrame, and will not change the original.`
`
* If you want to change the original DataFrame, use the `inplace = True argument:`

In [110]:
import pandas as pd

df = pd.read_csv('data1.csv')

# Remove all rows with NULL values:
df.dropna(inplace=True)

print(df.to_string())

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'  

### Replace Empty Values
* Another way of dealing with empty cells is to insert a new value instead.

* This way you do not have to delete entire rows just because of some empty cells.

* The `fillna() method `allows us to replace empty cells with a value:

In [115]:
import pandas as pd

df = pd.read_csv('data2.csv')
df.fillna(130, inplace = True)
print(df.to_string())

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'  

### Replace Using Mean, Median, or Mode
* A common way to replace empty cells, is to calculate the mean, median or mode value of the column.

* Pandas uses the `mean() median() and mode() methods `to calculate the respective values for a specified column:

In [122]:
import pandas as pd

df = pd.read_csv("data1.csv")
df.dropna (inplace=True)
x = df["Calories"].mean()
y = df["Calories"].median()
x = df["Calories"].mode()[0]
print(f"Mean of Calories: {x}")
print(f"Median of Calories: {y}")
print(f"Mode of Calories: {x}")
df.fillna({"Calories": x}, inplace=True)
df.fillna({"Duration": y}, inplace=True)




Mean of Calories: 300.0
Median of Calories: 291.2
Mode of Calories: 300.0


## Pandas - Cleaning Data of Wrong Format

In [124]:
import pandas as pd
df = pd.read_csv('data2.csv')
print(df.to_string())



    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'  

 ## Data of Wrong Format
* Cells with data of wrong format can make it difficult, or even impossible, to analyze data.

* To fix it, you have two options: `remove the rows`, or `convert all cells in the columns into the same format.`
* Pandas has a `to_datetime()` method for this:

In [125]:
import pandas as pd

df = pd.read_csv('data2.csv')

df['Date'] = pd.to_datetime(df['Date'], format='mixed')
print(df.to_string())

    Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90 

In [128]:
# Remove rows with a NULL value in the "Date" column:
df.dropna(subset=['Date'], inplace = True)
print(df.to_string())

    Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90 